### PR TITLE
Error if # is used in rpcpassword in conf

### LINKED ISSUE #13143: RPC Feature
I spent hours trying to solve this bug and the solution was so simple. 
My rpcpassword in my bitcoin.conf file contained a '#' character.

Prevent this from happening, by outputting this an error or warning when users run bitcoind.

Right now, you can have a # character as a password and it will work on the command line but will fail when you JSON-RPC.

### PR DESCRIPTION
**Pull request by meshcollider**
Fixes #13143 now #13482 was merged

### CODE REVIEW

File: src/util.cpp [hunk 1]
```diff
@@ -840,6 +842,10 @@ static bool GetConfigOptions(std::istream& stream, std::string& error, std::vect
             } else if ((pos = str.find('=')) != std::string::npos) {
                 std::string name = prefix + TrimString(str.substr(0, pos), pattern);
                 std::string value = TrimString(str.substr(pos + 1), pattern);
+                if (used_hash && name == "rpcpassword") {
+                    error = strprintf("parse error on line %i, illegal hash character used in rpcpassword", linenr);
```
Comment (laanwj) [inline src/util.cpp]: concept ACK—
though please reword this message, the `#` is not so much *illegal*, it starts a legal comment; it's just that *probably* the user made a mistake, assuming the `#`  will actually be part of the password

Comment (MeshCollider) [reply src/util.cpp]: Fixed :)

File: src/util.cpp [hunk 2]
```diff
@@ -840,6 +842,10 @@ static bool GetConfigOptions(std::istream& stream, std::string& error, std::vect
             } else if ((pos = str.find('=')) != std::string::npos) {
                 std::string name = prefix + TrimString(str.substr(0, pos), pattern);
                 std::string value = TrimString(str.substr(pos + 1), pattern);
+                if (used_hash && name == "rpcpassword") {
+                    error = strprintf("parse error on line %i, using # in rpcpassword can be ambiguous and should be avoided", linenr);
```
Comment (hebasto) [inline src/util.cpp]: Could `#` be mentioned as "reserved character"?

**Review by promag — COMMENTED** [review]
utACK fa3569d, simple solution.

It's a bit awkward to have this limitation though. For instance :
```
#                  split here because of whitespace before of #
#                  v
rpcpassword=foo#bar # comment # more comments..

# allow quotes, however it should allow escaping quotes too
pcpassword="foo#bar\"bar"# comment
```

Maybe it should warn each comment in the middle of lines?

**Comment by MeshCollider** [comment]
I think allowing quotes makes the situation even more ambiguous and confusing, I mentioned it on IRC. The whitespace split seems ok but I'm not sure it's worth it, better safe than sorry?

**Comment by promag** [comment]
Maybe it should also error with command line `-rpcpassword=foo#bar` to make it consistent?

**Comment by hebasto** [comment]
There is another approach that allows `#` in the `rpcpassword` value:
```
static bool GetConfigOptions(std::istream& stream, std::string& error, std::vector<std::pair<std::string, std::string>> &options)
{
    std::string str, prefix;
    std::string::size_type pos;
    int linenr = 1;
    while (std::getline(stream, str)) {
        // A number sign (#) at the beginning of the line indicates a comment. Comment lines are ignored. 
        if ((pos = str.find('#')) != 0) {
            const static std::string pattern = " \t\r\n";
            str = TrimString(str, pattern);
            if (!str.empty()) {
                if (*str.begin() == '[' && *str.rbegin() == ']') {
                    prefix = str.substr(1, str.size() - 2) + '.';
                } else if (*str.begin() == '-') {
                    error = strprintf("parse error on line %i: %s, options in configuration file must be specified without leading -", linenr, str);
                    return false;
                } else if ((pos = str.find('=')) != std::string::npos) {
                    std::string name = prefix + TrimString(str.substr(0, pos), pattern);
                    std::string value = TrimString(str.substr(pos + 1), pattern);
                    options.emplace_back(name, value);
                } else {
                    error = strprintf("parse error on line %i: %s", linenr, str);
                    if (str.size() >= 2 && str.substr(0, 2) == "no") {
                        error += strprintf(", if you intended to specify a negated option, use %s=1 instead", str);
                    }
                    return false;
                }
            }
        }
        ++linenr;
    }
    return true;
}
```

**Comment by promag** [comment]
@hebasto unfortunately that is breaking change.

**Comment by ryanofsky** [comment]
utACK fa3569dbf16bc40c36f036e191a9de8a679d81f8. Erroring on `#` does seem like the simplest way to resolve this issue, if `rpcpassword` is considered deprecated in favor of `rpcauth` anyway.

If we ever decide we need to allow `#` values in the future, it seems like it would be straightforward to do this in a mostly-backwards compatible way by supporting simple quoting like `name="value!@#$%^" # comment`.

**Review by ryanofsky — APPROVED** [review]

**Comment by MeshCollider** [comment]
Rebased

**Comment by MarcoFalke** [comment]
utACK 0385109444646561a718f34ae437b7e0e4d4d5bc

**Pull request merged**